### PR TITLE
Allows to set list of moves for botany play client

### PR DESCRIPTION
**Pull request by percevalve**
It takes into account the odd and even number of moves.
Uses the rerun_game from runner.py .
It does not take into account the state of the bot when doing the rerun-game.

### CODE REVIEW

File: client/src/botany_client/client.py [hunk 1]
```diff
@@ -147,6 +148,16 @@ def wrapped(board, move_list, token, state):
     if opcode_limit == 0:
         opcode_limit = None
 
+    #Initiate the move list parameter
+    if move_list is None:
+        move_list = []
+    else:
+        assert move_list.isdigit(), "move-list should only consist of integers"
+        move_list = list(move_list)
+        move_list = [int(move) for move in move_list]
```
Comment (inglesp) [inline client/src/botany_client/client.py]: You could combine these two lines into one:

```
move_list = [int(move) for move in move_list]
```

File: client/src/botany_client/client.py [hunk 2]
```diff
@@ -147,6 +148,16 @@ def wrapped(board, move_list, token, state):
     if opcode_limit == 0:
         opcode_limit = None
 
+    #Initiate the move list parameter
+    if move_list is None:
+        move_list = []
+    else:
+        assert move_list.isdigit(), "move-list should only consist of integers"
```
Comment (inglesp) [inline client/src/botany_client/client.py]: We don't want this `assert`, as it is specific to the rules in `botany_connectfour.game`, and other games might allow non-digits as moves.

File: client/src/botany_client/client.py [hunk 3]
```diff
@@ -147,6 +148,16 @@ def wrapped(board, move_list, token, state):
...
+        move_list = list(move_list)
+        move_list = [int(move) for move in move_list]
+        assert all(a < game.NCOLS for a in move_list ), f"Move cannot be greater then {game.NCOLS-1}"
```
Comment (inglesp) [inline client/src/botany_client/client.py]: This `assert` is also wrong, as it is again specific to `botany_connectfour.game`.

File: core/src/botany_core/runner.py [hunk 4]
```diff
@@ -44,7 +44,7 @@ def rerun_game(game, move_list):
     return boards
 
 
-def run_game(game, fn1, fn2, opcode_limit=None, display_board=False):
+def run_game(game, fn1, fn2, opcode_limit=None, display_board=False,move_list=[]):
```
Comment (inglesp) [inline core/src/botany_core/runner.py]: Using a list as a default argument is almost never what you want in Python.  [See this article](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) for more.

File: core/src/botany_core/runner.py [hunk 5]
```diff
@@ -61,13 +61,23 @@ def build_result(result_type, score, traceback=None, invalid_move=None):
     states = [None, None]
     winning_scores = [1, -1]
     losing_scores = [-1, 1]
-    board = game.new_board()
-    move_list = []
+    #Create a variable for cycle to allow rerun of odd number of moves
+    the_cycle = [0, 1]
```
Comment (inglesp) [inline core/src/botany_core/runner.py]: `the_cycle` doesn't tell me very much about what this variable is for.  Could you rename to `player_ixs`?

File: core/src/botany_core/runner.py (hunk 4 above)
Comment (percevalve) [reply core/src/botany_core/runner.py]: @inglesp Thanks for the link!

File: client/src/botany_client/client.py (hunk 3 above)
Comment (percevalve) [reply client/src/botany_client/client.py]: Was thinking at ConnectFour only.

**Comment by percevalve** [comment]
Hi @inglesp, 
Thanks for the review, should have resolved all comments.
For the state of the bots, to solve it, I added 2 additional option on the command line called last-state-bot1 and last-state-bot2. This the same way as this is solved in play as in views .py.

File: core/src/botany_core/runner.py (hunk 5 above)
Comment (percevalve) [reply core/src/botany_core/runner.py]: @inglesp Yes indeed. Updated it.

**Comment by inglesp** [comment]
Hi @percevalve, sorry for the delay in getting back to you on this.  The state parameters that you pass through to `run_game` will only work if the state is a string.  I'd like to get this PR merged, so please could you (a) remove them and (b) add a note to the help text for `--moves` to say that it'll only work if neither bot expects state.  We should then do some thinking about the best way to pass in state.

**Comment by percevalve** [comment]
HI @inglesp, thanks for the review, I was also a little busy on my side. Indeed this makes more sense.
I took out the funny way I created the variable in the runner function to trigger a message only if a list of moves was specified...
Instead, I just specified in the help that "List of moves, but only work if neither bot expects state.".

**Comment by inglesp** [comment]
Committed in 6e7955c0c944a6f66d9c590991c37800a40ff3db.  Thanks!

**Pull request closed without merging**